shared/reducers/login.tsx

```diff
@@ -1,28 +1,30 @@
-import * as Constants from '../constants/login'
 import * as Types from '../constants/types/login'
 import * as LoginGen from '../actions/login-gen'
 import * as SignupGen from '../actions/signup-gen'
 import * as ProvisionGen from '../actions/provision-gen'
+import * as Container from '../util/container'
 
-const initialState = Constants.makeState()
+const initialState = {
+  forgotPasswordSubmitting: false,
+  forgotPasswordSuccess: false,
+  registerUserPassLoading: false,
+}
+
+type Actions = LoginGen.Actions | SignupGen.RequestAutoInvitePayload | ProvisionGen.StartProvisionPayload
 
-export default function(
-  state: Types.State = initialState,
-  action: LoginGen.Actions | SignupGen.RequestAutoInvitePayload | ProvisionGen.StartProvisionPayload
-): Types.State {
-  switch (action.type) {
-    case LoginGen.resetStore:
-      return initialState
-    case SignupGen.requestAutoInvite: // fallthrough
-    case LoginGen.login: // fallthrough
-    case ProvisionGen.startProvision:
-      return state.merge({error: initialState.error})
-    case LoginGen.loginError:
-      return state.merge({error: action.payload.error || initialState.error})
-    case LoginGen.loadedIsOnline:
-      return state.merge({isOnline: action.payload.result})
-    // Saga only actions
-    default:
-      return state
-  }
+const clearErrors = (draftState: Container.Draft<Types.State>) => {
+  draftState.error = undefined
 }
+
+export default Container.makeReducer<Actions, Types.State>(initialState, {
+    [LoginGen.resetStore]: () => initialState,
+    [LoginGen.loginError]: (draftState, action) => {
+        draftState.error = action.payload.error
+    },
+    [LoginGen.loadedIsOnline]: (draftState, action) => {
+        draftState.isOnline = action.payload.isOnline
+    },
+    [SignupGen.requestAutoInvite]: (draftState) => clearErrors(draftState),
```

Comment:
why are these written as `(draftState) => clearErrors(draftState)` rather than just `clearErrors`?

Reply:
its a little dangerous to pass along closures like this, its very easy to get unexpected results so its something i avoid in general. else people will do things like passing closures to setTimeout or reduce or something and be surprised